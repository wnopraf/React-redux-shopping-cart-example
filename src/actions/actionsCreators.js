import { ADD_PRODUCT, DELETE_PRODUCT, INCREMENT_SOTCK, DECREMENT_SOTCK, INCREMENT_AMOUNT, DECREMENT_AMOUNT } from '../constants'

export const AddProduct = (id) => {
  return {
    type: ADD_PRODUCT,
    id
  }
}

export const deleteProduct = (id) => {
  return {
    type: DELETE_PRODUCT,
    id
  }
}

const incrementStock = (id) => ({ type: INCREMENT_SOTCK, id })

const decrementStock = (id) => ({ type: DECREMENT_SOTCK, id })

const incrementAmount = (id) => ({ type: INCREMENT_AMOUNT, id })

const decrementAmount = (id) => ({ type: DECREMENT_AMOUNT, id })

export const incrementAction = (dispatch, getState) => (id) => {
  dispatch(decrementStock(id))
  const { cart: { amount } } = getState()
  selectIdItem(amount, id).amount > 0 && dispatch(incrementAmount(id))
}
export const decrementAction = (dispatch, getState) => (id) => {
  dispatch(incrementStock(id))
  dispatch(decrementAmount(id))
}
const selectIdItem = (stateSlice, id) => stateSlice.find(e => e.id === id)
