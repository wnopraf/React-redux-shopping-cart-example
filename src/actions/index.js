import { INCREMENT_SOTCK, DECREMENT_SOTCK, INCREMENT_AMOUNT, DECREMENT_AMOUNT } from '../constants'

const incrementStock = (id) => ({ type: INCREMENT_SOTCK, id })

const decrementStock = (id) => ({ type: DECREMENT_SOTCK, id })

const incrementAmount = (id) => ({ type: INCREMENT_AMOUNT, id })

const decrementAmount = (id) => ({ type: DECREMENT_AMOUNT, id })

export const incrementAction = (dispatch, getState) => (id) => {
  const { stock } = getState()
  if (selectIdItem(stock, id).stock > 0) {
    dispatch(decrementStock(id))
    dispatch(incrementAmount(id))
  }
}
export const decrementAction = (dispatch, getState) => (id) => {
  dispatch(incrementStock(id))
  dispatch(decrementAmount(id))
}
const selectIdItem = (stateSlice, id) => stateSlice.find(e => e.productId === id)
