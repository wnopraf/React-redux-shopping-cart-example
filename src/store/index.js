import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import { stock, products } from '../data/products'
import cartReducer from '../reducers'

const store = createStore(cartReducer, { products, stock }, applyMiddleware(thunk))

export default store
