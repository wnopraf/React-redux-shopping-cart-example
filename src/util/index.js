export const selectIdItem = (stateSlice, id) => stateSlice.find(e => e.productId === id)
