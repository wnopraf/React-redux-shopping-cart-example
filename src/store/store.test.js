import store from '.'
import { decrementAction, incrementAction } from '../actions'
import { selectIdItem } from '../util'

describe('How store behaves on actions', () => {
  test.skip('First store snap', () => {
    expect(store.getState()).toMatchSnapshot()
  })
  const { cart, stock: initialStock } = store.getState()
  describe('Adding product to cart', () => {
    it('Cart is an Array', () => {
      store.dispatch(incrementAction(7))
      expect(cart).toBeInstanceOf(Array)
    })
    it('Cart has one item', () => {
      expect(store.getState().cart).toHaveLength(1)
    })
    it('Stock has one less unit', () => {
      const afterStockItem = selectIdItem(store.getState().stock, 7)
      const beforeStockItem = selectIdItem(initialStock, 7)
      expect(beforeStockItem.stock).toBe(afterStockItem.stock + 1)
    })
    it('First item check', () => {
      expect(store.getState().cart[0]).toEqual({ productId: 7, amount: 1 })
    })

    it('Cart has two item', () => {
      store.dispatch(incrementAction(2))
      expect(store.getState().cart).toHaveLength(2)
    })
    it('Second item check', () => {
      expect(store.getState().cart[1]).toEqual({ productId: 2, amount: 1 })
    })
    it('Second item stock has one less unit', () => {
      const afterStockItem = selectIdItem(store.getState().stock, 2)
      const beforeStockItem = selectIdItem(initialStock, 2)
      expect(beforeStockItem.stock).toBe(afterStockItem.stock + 1)
    })
    it('Adding three more items to cart', () => {
      for (let index = 5; index < 8; index++) {
        store.dispatch(incrementAction(index))
      }
      expect(store.getState().cart).toHaveLength(4)
    })
    it('Item 7 has amount 2', () => {
      const item = selectIdItem(store.getState().cart, 7)
      expect(item.amount).toBe(2)
    })
  })
  describe('Deleting product of cart', () => {
    it('Cart has 3 items', () => {
      store.dispatch(decrementAction(2))
      expect(store.getState().cart).toHaveLength(3)
    })
    it('Item 2 stock has been upgraded correctly', () => {
      const LastState = selectIdItem(store.getState().stock, 2)
      const defaultSate = selectIdItem(initialStock, 2)
      expect(LastState).toEqual(defaultSate)
    })
    it('Checking amount 0 validation', () => {
      store.dispatch(decrementAction(2))
      const { cart: nowCart, stock } = store.getState()
      const cartItem = selectIdItem(nowCart, 2)
      const stockItem = selectIdItem(stock, 2)
      const defaultStockItem = selectIdItem(initialStock, 2)
      expect.assertions(2)
      expect(cartItem).toBe(undefined)
      expect(stockItem).toEqual(defaultStockItem)
    })
  })
})
