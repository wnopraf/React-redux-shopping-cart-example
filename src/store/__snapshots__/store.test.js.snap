// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`How store behaves on actions First store snap 1`] = `
Object {
  "amount": Array [],
  "products": Array [
    Object {
      "id": 1,
      "name": "product-A",
      "prize": "147.17",
    },
    Object {
      "id": 2,
      "name": "product-B",
      "prize": "91.78",
    },
    Object {
      "id": 3,
      "name": "product-C",
      "prize": "139.51",
    },
    Object {
      "id": 4,
      "name": "product-D",
      "prize": "76.43",
    },
    Object {
      "id": 5,
      "name": "product-E",
      "prize": "99.69",
    },
    Object {
      "id": 6,
      "name": "product-F",
      "prize": "124.42",
    },
    Object {
      "id": 7,
      "name": "product-G",
      "prize": "86.04",
    },
    Object {
      "id": 8,
      "name": "product-H",
      "prize": "121.05",
    },
    Object {
      "id": 9,
      "name": "product-I",
      "prize": "130.44",
    },
    Object {
      "id": 10,
      "name": "product-J",
      "prize": "90.88",
    },
    Object {
      "id": 11,
      "name": "product-K",
      "prize": "94.98",
    },
    Object {
      "id": 12,
      "name": "product-E",
      "prize": "90.93",
    },
    Object {
      "id": 13,
      "name": "product-L",
      "prize": "138.89",
    },
    Object {
      "id": 14,
      "name": "product-M",
      "prize": "99.31",
    },
    Object {
      "id": 15,
      "name": "product-N",
      "prize": "84.24",
    },
    Object {
      "id": 16,
      "name": "product-Ã‘",
      "prize": "111.76",
    },
    Object {
      "id": 17,
      "name": "product-O",
      "prize": "132.29",
    },
    Object {
      "id": 18,
      "name": "product-P",
      "prize": "84.03",
    },
    Object {
      "id": 19,
      "name": "product-Q",
      "prize": "115.00",
    },
    Object {
      "id": 20,
      "name": "product-R",
      "prize": "144.68",
    },
    Object {
      "id": 21,
      "name": "product-S",
      "prize": "77.10",
    },
    Object {
      "id": 22,
      "name": "product-T",
      "prize": "99.03",
    },
    Object {
      "id": 23,
      "name": "product-U",
      "prize": "92.04",
    },
    Object {
      "id": 24,
      "name": "product-V",
      "prize": "93.25",
    },
    Object {
      "id": 25,
      "name": "product-W",
      "prize": "106.48",
    },
    Object {
      "id": 26,
      "name": "product-Y",
      "prize": "130.89",
    },
    Object {
      "id": 27,
      "name": "product-Z",
      "prize": "60.40",
    },
  ],
  "stock": Array [
    Object {
      "productId": 1,
      "stock": 13,
    },
    Object {
      "productId": 2,
      "stock": 15,
    },
    Object {
      "productId": 3,
      "stock": 11,
    },
    Object {
      "productId": 4,
      "stock": 11,
    },
    Object {
      "productId": 5,
      "stock": 11,
    },
    Object {
      "productId": 6,
      "stock": 14,
    },
    Object {
      "productId": 7,
      "stock": 9,
    },
    Object {
      "productId": 8,
      "stock": 11,
    },
    Object {
      "productId": 9,
      "stock": 14,
    },
    Object {
      "productId": 10,
      "stock": 11,
    },
    Object {
      "productId": 11,
      "stock": 14,
    },
    Object {
      "productId": 12,
      "stock": 9,
    },
    Object {
      "productId": 13,
      "stock": 13,
    },
    Object {
      "productId": 14,
      "stock": 13,
    },
    Object {
      "productId": 15,
      "stock": 14,
    },
    Object {
      "productId": 16,
      "stock": 8,
    },
    Object {
      "productId": 17,
      "stock": 11,
    },
    Object {
      "productId": 18,
      "stock": 13,
    },
    Object {
      "productId": 19,
      "stock": 11,
    },
    Object {
      "productId": 20,
      "stock": 14,
    },
    Object {
      "productId": 21,
      "stock": 13,
    },
    Object {
      "productId": 22,
      "stock": 10,
    },
    Object {
      "productId": 23,
      "stock": 12,
    },
    Object {
      "productId": 24,
      "stock": 15,
    },
    Object {
      "productId": 25,
      "stock": 12,
    },
    Object {
      "productId": 26,
      "stock": 14,
    },
    Object {
      "productId": 27,
      "stock": 14,
    },
  ],
}
`;
